### PR TITLE
Save reward and discriminator networks, and other improved logging

### PR DESCRIPTION
**Pull request by AdamGleave**
Currently we can train reward networks but have no way to persist them, which makes them of somewhat limited use. This adds a `save` and `load` method to `RewardNet` and `DiscrimNet`. Additionally, I add logic to `scripts/train.py` to save checkpoints and the final model. This required making some new logging logic, so I also added some logging for Stable Baselines and generally tidied up. This doesn't touch logging of statistics for the *discriminator* (which is still via matplotlib PNGs, and should be amended in another PR).

Concretely, I created a new ABC `Serializable` with abstract `save` and `load` methods, with a concrete implementation `LayersSerializable`. This saves the constructor arguments, needed to reconstruct the graph, and uses `tf.train.Checkpoint` to save/restore the parameter variables.

I'm not that happy with the current implementation. The main bugbears are: (a) mix-in for this is potentially overkill, and (b) has no way of serializing mutable Python state. It really feels like this is something that TensorFlow or a library built on top of it should be taking care of. But I've not been able to find anything. Keras makes this easy, but it's too limited for our purposes.

One alternative would be use `tf.train.Saver` instead of `tf.train.Checkpoint`. This would have the advantage that the user could would not need to keep track of TensorFlow state (instead, we just save all variables within a scope). However, TF 2.x removes the notion of variable collections, and `Saver` is deprecated for this reason. So it gives a cleaner upgrade path.

The other major bugbear is in serializing recursive structures. What I've done for `DiscrimNetAIRL` is really hacky: I save the wrapped reward network in a subdirectory!

If anyone can think of a better way of doing this I'd be very happy to hear it.

### CODE REVIEW

**Comment by shwang** [comment]
Seems fine to push after these changes. I'm surprised that the diff hit is so low right now (look into it myself, but I'm going to sleep soon)

File: imitation/util/serialize.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  def save_parameters(self, directory: str) -> None:
+    self._checkpoint.save(file_prefix=os.path.join(directory, "weights"))
+
+  def load_parameters(self, directory: str) -> None:
+    restore = self._checkpoint.restore(tf.train.latest_checkpoint(directory))
+    restore.assert_consumed().run_restore_ops()
+
+  @classmethod
+  def load(cls, directory: str) -> None:
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply imitation/util/serialize.py]: Huh, it is odd this passed type checking, thanks!

**Comment by AdamGleave** [comment]
Thanks for the comments as usual.

codecov got confused by the use of multiprocessing. I've disabled this in the codecov tests, and am mostly just testing with `DummyVecEnv`, keeping a few tests for `SubprocVecEnv` to check nothing is seriously broken. We're now back up to ~77% diff coverage (much higher outside of auxiliary), which is healthier.

Merging now.
